akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +997,176 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
+    private case object Initialized extends State
+    //stage has been shut down, either regularly or it failed
+    private case object Completed extends State
+    //Event with feedback promise
+    private case class Event(e: T, handlingPromise: OptionVal[Promise[Done]])
+
+    val concurrentSet = ConcurrentHashMap.newKeySet[Promise[_]]()
+
+    private[this] val currentState = new AtomicReference[State](Pending(List.empty[Event]))
+
+    //is called from stream - thread pool is managed by Akka
+    @tailrec
+    final private[stage] def onStart(): Unit = {
+      (currentState.getAndSet(Initializing): @unchecked) match {
+        case Pending(l) ⇒ l.reverse.foreach(ack ⇒ {
+          onAsyncInput(ack.e)
+        })
+      }
+      if (!currentState.compareAndSet(Initializing, Initialized)) {
+        (currentState.get: @unchecked) match {
+          case Pending(_) ⇒ onStart()
+          case Completed  ⇒ () //wonder if this is possible
+        }
+      }
+    }
+
+    //is called from stream - thread pool is managed by Akka
+    final private[stage] def onStop(): Unit = {
+      currentState.set(Completed)
+      concurrentSet.asScala.foreach(_.tryFailure(new StreamDetachedException()))
+      concurrentSet.clear()
+    }
+
+    private def onAsyncInput(event: T) = interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+
+    private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
+      onAsyncInput(event)
+      currentState.get() match {
+        case Completed ⇒ failPromiseOnComplete(promise)
+        case _         ⇒ promise
+      }
+    }
+
+    private def createFeedback(): Promise[Done] = {
+      val promise: Promise[Done] = Promise[Done]()
+      concurrentSet.add(promise)
+      promise.future.andThen { case _ ⇒ concurrentSet.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+      promise
+    }
+
+    //external call
+    override def invokeWithFeedback(event: T): Future[Done] = invokeWithPromise(event).result.future
+
+    private def invokeWithPromise(event: T): TailRec[Promise[Done]] = {
+      val promise: Promise[Done] = createFeedback
+
+      currentState.get() match {
+        //not started yet
+        case list @ Pending(_) ⇒ tailcall(onPending(list, promise, event))
+        //started - can just send message to stream
+        case Initialized       ⇒ done(sendEvent(event, promise))
+        //initializing is in progress in another thread (initializing thread is managed by Akka)
+        case Initializing ⇒ if (!currentState.compareAndSet(Initializing, Pending(Event(event, OptionVal(promise)) :: List.empty[Event]))) {
+          (currentState.get(): @unchecked) match {
+            case list @ Pending(_) ⇒ tailcall(onPending(list, promise, event))
+            case Initialized       ⇒ done(sendEvent(event, promise))
+          }
+        } else done(promise)
```

Comment:
Nitpick: A bit inconsistent that the other cases delegates to methods but this one does the recursion inline.